analyzers/src/SonarAnalyzer.CSharp/Rules/LoggingArgumentsShouldBePassedCorrectly.cs

```diff
@@ -0,0 +1,123 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using static Roslyn.Utilities.SonarAnalyzer.Shared.LoggingFrameworkMethods;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class LoggingArgumentsShouldBePassedCorrectly : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6668";
+    private const string MessageFormat = "Logging arguments should be passed to the correct parameter.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var invocation = (InvocationExpressionSyntax)c.Node;
+                if (!HasLoggingMethodName(invocation)
+                    || c.SemanticModel.GetSymbolInfo(invocation).Symbol is not IMethodSymbol invocationSymbol)
+                {
+                    return;
+                }
+                if (invocationSymbol.HasContainingType(KnownType.Microsoft_Extensions_Logging_LoggerExtensions, false))
```

Comment:
Restructuring: checking for every method name, logging framework, and parameter type separately seems unnecessary. It results in a lot of if-else if-else if... constructs. I don't think they need to be checked on a framework-by-framework basis.
[In one of the other logging rules](https://github.com/SonarSource/sonar-dotnet/blob/b0d5ee026c896a4cb61d990d358b0842280c2843/analyzers/src/SonarAnalyzer.CSharp/Rules/UseConstantLoggingTemplate.cs#L41) I implemented the check this way:
- check if the method name matches any of the logging method names (from any of the logging frameworks, it's just one big collection)
- then check if the caller is a logger object (also from any of the logging frameworks)
- then check for parameters: in that rule, it's done by parameter name, here it can be done by parameter type
Each step is a simple call to a `Contains/Any/FirstOrDefault` LINQ method.